File: domain/src/main/java/com/beat/domain/booking/dao/BookingRepository.java
package com.beat.domain.booking.dao;

import java.util.List;
import java.util.Optional;

import com.beat.domain.booking.domain.Booking;
import com.beat.domain.booking.domain.BookingStatus;

public interface BookingRepository {

	Booking save(Booking booking);

	Optional<Booking> findById(Long id);

	List<Booking> findAll();

	Optional<List<Booking>> findByBookerNameAndBookerPhoneNumberAndPasswordAndBirthDate(
		String bookerName,
		String bookerPhoneNumber,
		String password,
		String birthDate
	);

	Optional<Booking> findFirstByBookerNameAndBookerPhoneNumberAndBirthDateAndPassword(
		String bookerName,
		String bookerPhoneNumber,
		String birthDate,
		String password
	);

	List<Booking> findByUserId(Long userId);

	boolean existsActiveBookingByScheduleIds(
		List<Long> scheduleIds,
		List<BookingStatus> excludedStatuses
	);

	int deleteInactiveBookingsByScheduleIds(
		List<Long> scheduleIds,
		List<BookingStatus> inactiveStatuses
	);
}

File: domain/src/main/kotlin/com/beat/domain/booking/repository/BookingRepository.kt
package com.beat.domain.booking.repository

import com.beat.domain.booking.model.Booking
import com.beat.domain.booking.model.BookingStatus
import java.time.LocalDateTime

interface BookingRepository {
    fun save(booking: Booking): Booking

    fun findById(id: Long): Booking?

    fun findScheduleIdsByIds(ids: Collection<Long>): List<Long>

    fun lockById(id: Long): Booking?

    fun findAll(): List<Booking>

    fun deleteAll(bookings: Iterable<Booking>)

    fun findByBookingStatusAndCancellationDateBefore(
        bookingStatus: BookingStatus,
        cancellationDate: LocalDateTime,
    ): List<Booking>

    fun findByUserId(userId: Long): List<Booking>

    fun existsActiveBookingByScheduleIds(
        scheduleIds: List<Long>,
        excludedStatuses: List<BookingStatus>,
    ): Boolean

    fun deleteInactiveBookingsByScheduleIds(
        scheduleIds: List<Long>,
        inactiveStatuses: List<BookingStatus>,
    ): Int
}
